### PR TITLE
[sleepiq] Add functionality to control the bed foundation

### PR DESCRIPTION
**Pull request by mhilbush**
Adds new channels for controlling the bed foundation. Supports 
- adjusting the head and foot by selecting a preset
- adjusting the head and foot independently using a specific adjustment setting (between 0 and 100)
- controlling the night light outlet
- controlling the under-bed lighting

Channels are updated based on status returned form the SleepIQ API so it should stay in sync with operations performed using the smartphone app.

Signed-off-by: Mark Hilbush <mark@hilbush.com>

### CODE REVIEW

File: bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java [hunk 1]
```diff
@@ -262,6 +274,116 @@ public void setPauseMode(@Nullable String bedId, boolean mode) {
...
+     *
+     * @param bedId the bed identifier
+     * @param side the side of the bed
+     * @param actuator the head or foot of the bed
+     * @param position the new position of the actuator
+     * @param speed the speed with which to make the adjustment
+     */
+    public void setFoundationPosition(@Nullable String bedId, Side side, FoundationActuator actuator, int position,
+            FoundationActuatorSpeed speed) {
+        if (bedId == null) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java:350]: It doesn't make sense to call this method with `bedId == null`. Consider making it non-nullable and avoid calling it in this case.

File: bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java [hunk 2]
```diff
@@ -262,6 +274,116 @@ public void setPauseMode(@Nullable String bedId, boolean mode) {
...
+     * Set a foundation adjustment preset.
+     *
+     * @param bedId the bed identifier
+     * @param side the side of the bed
+     * @param preset the preset to be applied
+     * @param speed the speed with which to make the adjustment
+     */
+    public void setFoundationPreset(@Nullable String bedId, Side side, FoundationPreset preset,
+            FoundationActuatorSpeed speed) {
+        if (bedId == null) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java:325]: It doesn't make sense to call this method with `bedId == null`. Consider making it non-nullable and avoid calling it in this case.

File: bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java [hunk 3]
```diff
@@ -262,6 +274,116 @@ public void setPauseMode(@Nullable String bedId, boolean mode) {
...
+        return null;
+    }
+
+    /**
+     * Get the foundation status of the specified bed
+     *
+     * @param bedId the bed identifier
+     */
+    public @Nullable FoundationStatusResponse getFoundationStatus(@Nullable String bedId) {
+        if (bedId == null) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java:302]: It doesn't make sense to call this method with `bedId == null`. Consider making it non-nullable and avoid calling it in this case.

File: bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java [hunk 4]
```diff
@@ -262,6 +274,116 @@ public void setPauseMode(@Nullable String bedId, boolean mode) {
         }
     }
 
+    /**
+     * Get the foundation features of the specified bed
+     *
+     * @param bedId the bed identifier
+     */
+    public @Nullable FoundationFeaturesResponse getFoundationFeatures(@Nullable String bedId) {
+        if (bedId == null) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java:283]: It doesn't make sense to call this method with `bedId == null`. Consider making it non-nullable and avoid calling it in this case.

**Review by jlaur — CHANGES_REQUESTED** [review]
The code looks good and clean to me. I have added only a minor comment. Also, consider dynamically removing the new channels if a foundation is not present - see for example #13275.

File: bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java (hunk 2 above)
Comment (mhilbush) [reply bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java]: Fixed

File: bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java (hunk 3 above)
Comment (mhilbush) [reply bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java]: Fixed

File: bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java (hunk 4 above)
Comment (mhilbush) [reply bundles/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/internal/handler/SleepIQCloudHandler.java]: Fixed

**Review by jlaur — APPROVED** [review]
LGTM

**Comment by jlaur** [comment]
> I did think about making the foundation channels dynamic. I don't want to just remove them dynamically because they could be removed incorrectly if there's a transient error on the call to the SleepIQ API. Then you're kinda screwed until you delete and readd the thing, right?

Yes, it would need to make that decision only when there is a reliable response with the correct capabilities. There is nothing wrong with providing channels not supported by all devices (beds), so this would only be sugar on the top. 🙂

**Comment by jlaur** [comment]
> Yes, it would need to make that decision only when there is a reliable response with the correct capabilities. There is nothing wrong with providing channels not supported by all devices (beds), so this would only be sugar on the top. 🙂

Sorry, was leaving the computer, so missed to ask the implicit question: Do you want to do more in this PR, or prefer to merge as-is?

**Comment by mhilbush** [comment]
If you're good with it, merge as is please.

**Pull request merged**